platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/isolation/IsolatedActionSerializer.kt

```diff
@@ -0,0 +1,172 @@
+import java.io.OutputStream
+import java.util.IdentityHashMap
+
+
+/**
+ * Serialized state of an object graph containing one or more [IsolatedAction]s.
+ */
+class SerializedAction(
+    /**
+     * The serialized graph.
+     */
+    val graph: ByteArray,
+
+    /**
+     * External references that are not serialized directly as part of the [graph].
+     * These include [Class] and [BuildServiceProvider] references.
+     */
+    val environment: Map<Int, Any>,
+)
+
+
+internal
+class IsolatedActionSerializer(
+    private val owner: IsolateOwner,
+    private val beanStateWriterLookup: BeanStateWriterLookup,
+    private val isolatedActionCodecs: IsolatedActionCodecsFactory
+) {
+    fun serialize(action: Any): SerializedAction {
+        val outputStream = ByteArrayOutputStream()
+        val classCollector = ClassCollector()
+        serializeTo(outputStream, classCollector, action)
+        return SerializedAction(
+            outputStream.toByteArray(),
+            classCollector.toLookup()
+        )
+    }
+
+    private
+    fun serializeTo(outputStream: ByteArrayOutputStream, classCollector: ClassCollector, action: Any) {
+        writeContextFor(outputStream, classCollector).useToRun {
+            runWriteOperation {
+                withIsolate(owner) {
+                    write(action)
+                }
+            }
+        }
+    }
+
+    private
+    fun writeContextFor(
+        outputStream: OutputStream,
+        classEncoder: ClassEncoder,
+    ) = DefaultWriteContext(
+        codec = isolatedActionCodecs.isolatedActionCodecs(),
+        encoder = KryoBackedEncoder(outputStream),
+        beanStateWriterLookup = beanStateWriterLookup,
+        logger = logger,
+        tracer = null,
+        problemsListener = ThrowingProblemsListener,
+        classEncoder = classEncoder
+    )
+}
+
+
+internal
+class IsolatedActionDeserializer(
+    private val owner: IsolateOwner,
+    private val beanStateReaderLookup: BeanStateReaderLookup,
+    private val isolatedActionCodecs: IsolatedActionCodecsFactory
+) {
+    fun deserialize(action: SerializedAction): Any =
+        readerContextFor(action).useToRun {
+            runReadOperation {
+                withIsolate(owner) {
+                    readNonNull()
+                }
+            }
+        }
+
+    private
+    fun readerContextFor(action: SerializedAction) = DefaultReadContext(
+        codec = isolatedActionCodecs.isolatedActionCodecs(),
+        decoder = KryoBackedDecoder(action.graph.inputStream()),
+        beanStateReaderLookup = beanStateReaderLookup,
+        logger = logger,
+        problemsListener = ThrowingProblemsListener,
+        classDecoder = object : ClassDecoder {
+            override fun ReadContext.decodeClass(): Class<*> =
+                action.environment[readSmallInt()]?.uncheckedCast()!!
+        }
+    )
+}
+
+
+private
+class ClassCollector : ClassEncoder {
+
+    private
+    val classes = IdentityHashMap<Class<*>, Int>()
```

Comment:
I want to verify my understanding:
`IHM` is using referential equality check for the keys instead of structural equality check.
`Class` instances effectively are singletones within single classloader.
Because of each build script has own classloader, we want to have in this cache, say, Foo::class of `:a` buildscript is different from Foo::class of `:b` buildscript.
Is it correct?

Reply:
Yes, that's right. The identity of a `Class` includes its `ClassLoader`. In practice, equality would work just as well but it's a standard idiom to use `IdentityHashMap` in these cases.

Reply:
And I should mention, the long-term plan is that more references would be captured in this `environment` map so identity is the right choice because of that.